### PR TITLE
[Azure.Monitor.Ingestion] Restore Protocol methods to public

### PR DESCRIPTION
**Pull request by nisha-bhatia**
# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Comment by nisha-bhatia** [comment]
Chatted with LLC folks - we will GA with public Protocol (DPG) methods

File: sdk/monitor/Azure.Monitor.Ingestion/tests/MonitorIngestionLiveTest.cs [hunk 1]
```diff
@@ -175,5 +175,33 @@ public void ConcurrencySingleThread()
             Assert.AreEqual(UploadLogsStatus.Success, response.Value.Status);
             Assert.IsEmpty(response.Value.Errors);
         }
+
+        [Test]
+        public async Task ValidInputFromObjectAsJsonNoBatchingAsync()
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/tests/MonitorIngestionLiveTest.cs:180]: Do we also need a test that passes in some other content-encoding now that we are exposing that parameter?

File: sdk/monitor/Azure.Monitor.Ingestion/src/Generated/LogsIngestionClient.cs [hunk 2]
```diff
@@ -153,7 +153,7 @@ internal virtual async Task<Response> UploadAsync(string ruleId, string streamNa
         /// ]]></code>
         /// </example>
         /// <remarks> See error response code and error response message for more detail. </remarks>
-        internal virtual Response Upload(string ruleId, string streamName, RequestContent content, string contentEncoding = "gzip", RequestContext context = null)
+        public virtual Response Upload(string ruleId, string streamName, RequestContent content, string contentEncoding = "gzip", RequestContext context = null)
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/src/Generated/LogsIngestionClient.cs:156]: Do we Gzip the user content here? Or is this meant to be a method that will just take exactly what the user provides and send it to the service. If that is the case, I'm not sure that it makes sense to default the contentEncoding parameter as "gzip".

File: sdk/monitor/Azure.Monitor.Ingestion/CHANGELOG.md [hunk 3]
```diff
@@ -1,14 +1,8 @@
...
 ## 1.0.0-beta.4 (Unreleased)
-
-### Features Added
-
-### Breaking Changes
-
-### Bugs Fixed
-
-### Other Changes
+## Features Added
```
Comment (jsquire) [inline sdk/monitor/Azure.Monitor.Ingestion/CHANGELOG.md:4]:
```suggestion

## Features Added
```

File: sdk/monitor/ci.yml [hunk 4]
```diff
@@ -13,7 +13,7 @@ trigger:
     - sdk/monitor/service.projects
     - sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter
     - sdk/monitor/Azure.Monitor.Query
-    - sdk/monitor/AzureMonitorIngestion
+    - sdk/monitor/Azure.Monitor.Ingestion
```
Comment (jsquire) [inline sdk/monitor/ci.yml:16]: This is also needed on L31

File: sdk/monitor/Azure.Monitor.Ingestion/src/Generated/LogsIngestionClient.cs (hunk 2 above)
Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Ingestion/src/Generated/LogsIngestionClient.cs:156]: We are trying to gzip by default, so any input provided by the user will be gzipped

Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Ingestion/src/Generated/LogsIngestionClient.cs:156]: request gets constructed on 109

Comment (JoshLove-msft) [reply sdk/monitor/Azure.Monitor.Ingestion/src/Generated/LogsIngestionClient.cs:156]: Got it! So this method will still GZip for users based on the content-encoding.

File: sdk/monitor/Azure.Monitor.Ingestion/tests/MonitorIngestionLiveTest.cs [hunk 5]
```diff
@@ -39,7 +39,7 @@ private LogsIngestionClient CreateClient(HttpPipelinePolicy policy = null)
         public void NullInput()
         {
             LogsIngestionClient client = CreateClient();
-            Assert.ThrowsAsync<ArgumentNullException>(async () => await client.UploadAsync(TestEnvironment.DCRImmutableId, TestEnvironment.StreamName, null).ConfigureAwait(false));
+            Assert.ThrowsAsync<NullReferenceException>(async () => await client.UploadAsync(TestEnvironment.DCRImmutableId, TestEnvironment.StreamName, null).ConfigureAwait(false));
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/tests/MonitorIngestionLiveTest.cs:42]: We shouldn't need to change this test - this suggests that we are missing some null validation now that we are not calling the generated method.

File: sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs [hunk 6]
```diff
@@ -403,5 +403,115 @@ private static UploadLogsStatus GetStatus<T>(IEnumerable<T> logEntries, List<Upl
...
+        /// var data = new[] {
+        ///     new {}
+        /// };
+        ///
+        /// Response response = await client.UploadAsync("<ruleId>", "<streamName>", RequestContent.Create(data), <gzip>);
+        /// Console.WriteLine(response.Status);
+        /// ]]></code>
+        /// </example>
+        /// <remarks> See error response code and error response message for more detail. </remarks>
+        public virtual async Task<Response> UploadAsync(string ruleId, string streamName, RequestContent content, RequestContext context = null, CancellationToken cancellationToken = default)
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs:446]: Do we need to take in the CancellationToken as a parameter? I think it can be set on the RequestContext. We should try to match what the DPG generator does.

File: sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs [hunk 7]
```diff
@@ -15,6 +15,9 @@ namespace Azure.Monitor.Ingestion
 {
     /// <summary> The IngestionUsingDataCollectionRules service client. </summary>
     [CodeGenClient("IngestionUsingDataCollectionRulesClient")]
+    [CodeGenSuppress("Upload", typeof(string), typeof(string), typeof(RequestContent), typeof(string), typeof(RequestContext))]
+    [CodeGenSuppress("UploadAsync", typeof(string), typeof(string), typeof(RequestContent), typeof(string), typeof(RequestContext))]
+    [CodeGenSuppress("CreateUploadRequest", typeof(string), typeof(string), typeof(RequestContent), typeof(string), typeof(RequestContext))]
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs:20]: nit: I think this one is probably unnecessary as it shouldn't be included if the first two are excluded.

File: sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs (hunk 6 above)
Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs:449]: yes - it can be set in the RequestContext, I'll remove it from the DPG generated methods

**Review by jsquire — APPROVED** [review]

File: sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs [hunk 8]
```diff
@@ -403,5 +405,121 @@ private static UploadLogsStatus GetStatus<T>(IEnumerable<T> logEntries, List<Upl
...
+        /// var data = new[] {
+        ///     new {}
+        /// };
+        ///
+        /// Response response = await client.UploadAsync("<ruleId>", "<streamName>", RequestContent.Create(data), <gzip>);
+        /// Console.WriteLine(response.Status);
+        /// ]]></code>
+        /// </example>
+        /// <remarks> See error response code and error response message for more detail. </remarks>
+        public virtual async Task<Response> UploadAsync(string ruleId, string streamName, RequestContent content, RequestContext context = null)
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs:447]: nit: it might be worth creating a separate private method that has all the logic and having the sync/async public methods call into it.

File: sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs [hunk 9]
```diff
@@ -403,5 +403,121 @@ private static UploadLogsStatus GetStatus<T>(IEnumerable<T> logEntries, List<Upl
...
+        /// var data = new[] {
+        ///     new {}
+        /// };
+        ///
+        /// Response response = client.Upload("<ruleId>", "<streamName>", RequestContent.Create(data), <gzip>);
+        /// Console.WriteLine(response.Status);
+        /// ]]></code>
+        /// </example>
+        /// <remarks> See error response code and error response message for more detail. </remarks>
+        internal virtual Response Upload(string ruleId, string streamName, RequestContent content, RequestContext context = null)
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs:503]: Is this method used anywhere? Maybe this can be the internal method that the public methods call into.

Comment (nisha-bhatia) [reply sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs:503]: this method is supposed to be public

File: sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs [hunk 10]
```diff
@@ -403,5 +403,118 @@ private static UploadLogsStatus GetStatus<T>(IEnumerable<T> logEntries, List<Upl
...
+        /// };
+        ///
+        /// Response response = client.Upload("<ruleId>", "<streamName>", RequestContent.Create(data), <gzip>);
+        /// Console.WriteLine(response.Status);
+        /// ]]></code>
+        /// </example>
+        /// <remarks> See error response code and error response message for more detail. </remarks>
+        public virtual Response Upload(string ruleId, string streamName, RequestContent content, RequestContext context = null)
+        {
+            return UploadRequestContentAsync(ruleId, streamName, content, false, context).Result;
```
Comment (JoshLove-msft) [inline sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs:490]:
```suggestion
            return UploadRequestContentAsync(ruleId, streamName, content, false, context).EnsureCompleted();
```

**Review by JoshLove-msft — APPROVED** [review]
